Add Unity documentation (#592)
codemagic-ci-cd/codemagic-docs: content/yaml-quick-start/building-a-unity-app.md

---
title: Building Unity mobile apps
description: How to build Unity mobile apps with codemagic.yaml
weight: 12

---

Unity is a cross-platform game engine developed by Unity Technologies. It can be used to create mobile applications that run on iOS and Android.

Codemagic's Unity premium macOS base image has the Unity SDK, iOS and Android modules pre-installed. 

This guide will show you how to configure a workflow that builds and publishes your app to TestFlight and Google Play using a **codemagic.yaml** configuration file. 

{{<notebox>}}
**Important!** Building Unity apps on Codemagic is currently in closed beta release. If you are interested in taking part in the beta, please contact us [here](https://codemagic.io/contact/)
{{</notebox>}}

## Adding your Unity project to Codemagic

> The apps you have available on Codemagic are listed on the Applications page. Click **Add application** to add a new app.

Reviewer: Perhaps link to the relevant docs about adding new apps so as to skip this part here

1. On the Applications page, click **Set up build** next to the app you want to start building. 
2. On the popup, select **Other** as the project type and click **Continue**.

> 3. In your IDE, create a [`codemagic.yaml`](./yaml). This will be used to create our workflow.

Reviewer: The **Set up build** button and project type selection appear only for older repos. If the app is newly added, the project type selection happens already earlier in the flow, and on the Applications page the app shows up with **Finish build setup**. 
Actually the whole section here (and in other guides) is outdated. Instead, I would list here the requirements for building Unity (not necessarily in this order):
- Add a codemagic.yaml to the root of the repo and link to the workflow before
- Set up webhooks
- Add the Unity specific variables / activate the licence (link to the relevant section)
etc

4. Commit the configuration file to the root of your repository.
5. Back in app settings in Codemagic, scan for the `codemagic.yaml` file by selecting a **branch** to scan and clicking the **Check for configuration file** button at the top of the page. Note that you can have different configuration files in different branches.

## Environment variables for Unity

You will need to set the following environment variables for Unity specific values in Codemagic: 

`UNITY_HOME`, `UNITY_SERIAL`, `UNITY_USERNAME` and `UNITY_PASSWORD`.

> You can add these as global environment variables for your team as follows (make sure the secure option is checked for any sensitive values):

Reviewer: If it's possible to set up Unity only in teams, this should be listed as a requirement too

Codemagic's Unity premium macOS base image has the Unity SDK, iOS and Android modules pre-installed. 

This guide will show you how to configure a workflow that builds and publishes your app to TestFlight and Google Play using a **codemagic.yaml** configuration file. 

{{<notebox>}}
**Important!** Building Unity apps on Codemagic is currently in closed beta release. If you are interested in taking part in the beta, please contact us [here](https://codemagic.io/contact/)
{{</notebox>}}

## Adding your Unity project to Codemagic

It is possible to add repositories from Github, Gitlab, Bitbucket or any Git based repository. Please refer to the following documentation about adding apps from these sources.

- [Adding apps from GitHub](../getting-started/github/).
- [Adding apps from GitLab](../getting-started/gitlab/).
- [Adding apps from Bitbucket](../getting-started/bitbucket/).
- [Adding apps from other repositories](../getting-started/other/).

## Setting up your workflow

These are the key steps in setting up your workflow for building Unity mobile apps: 

> 1. On the Applications page in Codemagic click the 'Add application' button and follow the on-screen instructions to add your Unity project.

Reviewer: ```suggestion
1. On the Applications page in Codemagic, click the **Add application** button and follow the on-screen instructions to add your Unity project.
```

---
title: Building Unity mobile apps
description: How to build Unity mobile apps with codemagic.yaml
weight: 12

---

Unity is a cross-platform game engine developed by Unity Technologies. It can be used to create mobile applications that run on iOS and Android.

Codemagic's Unity premium macOS base image has the Unity SDK, iOS and Android modules pre-installed. 

This guide will show you how to configure a workflow that builds and publishes your app to TestFlight and Google Play using a **codemagic.yaml** configuration file. 

## Prerequisites

- A Unity **Plus** or **Pro** license. Your license is used to activate Unity on the Codemagic build server so the iOS and Android projects can be exported. The license is then returned before the workflow builds and publishes your apps to the App Store and Google Play.

> - If you are publishing to the Apple App Store you will need an active membership for the Apple Developer Program.

Reviewer: ```suggestion
- If you are publishing to the Apple App Store, you will need an active membership for the Apple Developer Program.
```

> - For publishing to Google Play you will require an active Google Play Developer account.

Reviewer: ```suggestion
- For publishing to Google Play, you will require an active Google Play Developer account.
```

3. Add environment variables for iOS code signing as described [here]({{< ref "#ios-variables" >}}).
4. Add environment variables for Android code signing as described [here]({{< ref "#android-variables" >}}).
5. Add a build script to Unity as explained [here]({{< ref "#unity-build-script" >}}).
6. Add a post-processing script to Unity for iOS builds as described [here]({{< ref "#post-processing-script" >}}).
7. Set the iOS bundle identifier in Unity as described [here]({{< ref "#bundle-identifier" >}}).
8. Configure Custom Gradle templates as described [here]({{< ref "#custom-gradle-templates" >}}).
9. Add a script for license activation, mobile app export and license return [here]({{< ref "#license-activation" >}}).
10. Create a codemagic.yaml workflow configuration using the Unity template and configure with your details as shown [here]({{< ref "#workflow-configuration" >}}).

## Environment variables for Unity {#unity-variables}

You will need to set the following environment variables for Unity specific values in Codemagic: 

`UNITY_HOME`, `UNITY_SERIAL`, `UNITY_USERNAME` and `UNITY_PASSWORD`.

> If you are using a Team in Codemagic, you can add these as global environment variables for your team by clicking on Teams > your Team name and then clicking on Global variables and secrets. Otherwise you can add the environment variables at application level by clicking the Environment variables tab.

Reviewer: ```suggestion
If you are using a Team in Codemagic, you can add these as global environment variables for your team by clicking on **Teams > your Team name** and then clicking on **Global variables and secrets**. Otherwise, you can add the environment variables at the application level by clicking the **Environment variables** tab.
```

> Add the environment variables as follows (make sure the secure option is checked for any sensitive values):

Reviewer: ```suggestion
Add the environment variables as follows (make sure the **Secure** option is checked for any sensitive values):
```

> 1. Create a variable called `UNITY_HOME` and set the value to `/Applications/Unity/Hub/Editor/2020.3.20f1/Unity.app` and click the 'Add' button.

Reviewer: ```suggestion
1. Create a variable called `UNITY_HOME` and set the value to `/Applications/Unity/Hub/Editor/2020.3.20f1/Unity.app` and click the **Add** button.
```

> 2. Create a variable called `UNITY_SERIAL` and set the value to your Unity serial number. In the 'Select group' dropdown type `unity` and click on the 'Create "unity" group' button. Mark the variable as **secure** to encrypt the value and click the 'Add' button.

Reviewer: ```suggestion
2. Create a variable called `UNITY_SERIAL` and set the value to your Unity serial number. In the **Select group** dropdown type `unity` and click on the **Create "unity" group** button. Mark the variable as **secure** to encrypt the value and click the **Add** button.
```

> 3. Create a variable called `UNITY_USERNAME` and set the value to **email address** used with your Unity ID , add to the "unity' group,  mark this as **secure** to encrypt the value and click the 'Add' button.

Reviewer: ```suggestion
3. Create a variable called `UNITY_USERNAME` and set the value to **email address** used with your Unity ID , add to the "unity" group,  mark this as **secure** to encrypt the value and click the **Add** button.
```

> 4. Create a variable called `UNITY_PASSWORD` and set the value to your Unity ID **password**, add to the "unity' group, mark this as **secure** to encrypt the value and click the 'Add' button.

Reviewer: ```suggestion
4. Create a variable called `UNITY_PASSWORD` and set the value to your Unity ID **password**, add to the "unity" group, mark this as **secure** to encrypt the value and click the **Add** button.
```

## Environment variables for iOS code signing {#ios-variables}

You will need to create the following environment variables in a group called `ios_credentials` for iOS code signing:

`APP_STORE_CONNECT_KEY_IDENTIFIER`, `APP_STORE_CONNECT_ISSUER_ID`, `APP_STORE_CONNECT_PRIVATE_KEY` and `CERTIFICATE_PRIVATE_KEY`

Please refer to the documentation about signing iOS apps [here](../yaml-code-signing/signing-ios/) for further details.

## Environment variables for Android code signing {#android-variables}

> You will need to set the following environment variables in a group called variable group called `keystore_credentials` for Android code signing:

Reviewer: ```suggestion
You will need to set the following environment variables in a variable group called `keystore_credentials` for Android code signing:
```

    {
        BuildPlayerOptions buildPlayerOptions = new BuildPlayerOptions();
        buildPlayerOptions.locationPathName = "android";
        buildPlayerOptions.target = BuildTarget.Android;
        buildPlayerOptions.options = BuildOptions.None;
        buildPlayerOptions.scenes = GetScenes();

        Debug.Log("Building Android");
        BuildPipeline.BuildPlayer(buildPlayerOptions);
        Debug.Log("Built Android");
    }

    [MenuItem("Build/Build Xcode")]
    public static void BuildXcode()
    {
        BuildPlayerOptions buildPlayerOptions = new BuildPlayerOptions();
        buildPlayerOptions.locationPathName = "ios";
        buildPlayerOptions.target = BuildTarget.iOS;
        buildPlayerOptions.options = BuildOptions.None;
        buildPlayerOptions.scenes = GetScenes();

        Debug.Log("Building iOS");
        BuildPipeline.BuildPlayer(buildPlayerOptions);
        Debug.Log("Built iOS");
    }

    private static string[] GetScenes()
    {
        return (from scene in EditorBuildSettings.scenes where scene.enabled select scene.path).ToArray();
    }

}
```

## Add a post-processing script to Unity {#post-processing-script}

> When publishing your app to TestFlight or the App Store you will be asked if your app uses encryption. 

Reviewer: ```suggestion
When publishing your app to TestFlight or the App Store, you will be asked if your app uses encryption. 
```

You can automate your answer to this question by setting the key `ITSAppUsesNonExemptEncryption` in your app's `Info.plist` file and set the value to `NO` if the app doesn't use encryption. 

> For more details about complying with encryption export regulations please see [here](https://developer.apple.com/documentation/security/complying_with_encryption_export_regulations).

Reviewer: ```suggestion
For more details about complying with encryption export regulations, please see [here](https://developer.apple.com/documentation/security/complying_with_encryption_export_regulations).
```

                    keyPassword System.getenv()['FCI_KEY_PASSWORD']
                } 
        }
    }

    lintOptions {
        abortOnError false
    }

    aaptOptions {
        noCompress = ['.ress', '.resource', '.obb'] + unityStreamingAssets.tokenize(', ')
        ignoreAssetsPattern = "!.svn:!.git:!.ds_store:!*.scc:.*:!CVS:!thumbs.db:!picasa.ini:!*~"
    }**PACKAGING_OPTIONS**
}**REPOSITORIES**
**IL_CPP_BUILD_SETUP**
**SOURCE_BUILD_SETUP**
**EXTERNAL_SOURCES**

```
## Add Gradle build files

When exporting the Android project to disk, Unity does not add the required gradle files which allow the project to be built from the command line. Therefore, you should add them mananually as follows:

Add a folder called **Gradle** to the root of your project. 

Add the files `gradlew`, `gradle.bat` and the `gradle/wrapper` directory which contains `gradle-wrapper.jar` and `gradle-wrapper.properties`. If you do not have access to these files from another project then they are available in the Unity sample project [here](https://github.com/codemagic-ci-cd/codemagic-sample-projects/tree/main/unity/unity-demo-project).

## License activation and return {#license-activation}

> Your Unity license needs to be activated on the Codemagic build server so the XCode project can be created. 

Reviewer: ```suggestion
Your Unity license needs to be activated on the Codemagic build server so the Xcode project can be created. 
```

> Once this is done the license can be returned and Codemagic can continue with building and publishing your iOS app. 

Reviewer: ```suggestion
Once this is done, the license can be returned and Codemagic can continue with building and publishing your iOS app. 
```

> The best way to perform this step is to run an external shell script, which means that even if the workflow itself fails your license will be returned. 

Reviewer: ```suggestion
The best way to perform this step is to run an external shell script, which means that even if the workflow itself fails, your license will be returned. 
```

Create the script as follows:

1. Open Terminal and run `touch export_unity.sh` to create a new shell script file. 

> 2. Before checking this file into source control run `chmod +x export_unity.sh` so the script can be executed.

Reviewer: ```suggestion
2. Before checking this file into source control, run `chmod +x export_unity.sh` so the script can be executed.
```

if [ $APP_TYPE = "ios" ]
then
  echo "UNITY BUILD IOS START"
  $UNITY_BIN -quit -batchmode -projectPath $UNITY_PROJECT_PATH -executeMethod BuildScript.BuildXcode -nographics -logfile $UNITY_LOG_FILE_IOS
  echo "UNITY BUILD IOS END"
fi

if [ $APP_TYPE = "android" ]
then
  echo "UNITY BUILD ANDROID START"
  $UNITY_BIN -quit -batchmode -projectPath $UNITY_PROJECT_PATH -executeMethod BuildScript.BuildAndroid -nographics -logfile $UNITY_LOG_FILE_ANDROID
  "UNITY BUILD ANDROID END"
fi

echo "UNITY RETURN LICENSE START"
$UNITY_BIN -quit -batchmode -returnlicense -nographics
echo "UNITY RETURN LICENSE END"

if [ $? -eq 0 ]; then
  echo "Unity Build Complete"
else
  echo "Unity Build Failed"
  exit 1;
fi
sleep 1

echo "Done everything"

exit 0
```

## Workflow configuration with codemagic.yaml{#workflow-configuration}

Your workflow should be configured using the **codemagic.yaml** configuration file and checked into the root of the branches you wish to build using Codemagic.

Add the following to your **codemagic.yaml** configuration file:

```yaml
workflows:
  unity-ios-workflow:
    name: Unity iOS Workflow
    max_build_duration: 120
    environment:
      groups:

>       # Add the group environment variables in Codemagic UI (either in Application/Team variables) - https://docs.codemagic.io/variables/environment-variable-groups/

Reviewer: ```suggestion
      # Add the group environment variables in Codemagic UI (either in Application or Team variables) - https://docs.codemagic.io/variables/environment-variable-groups/
```

          ./export_unity.sh ios
      - name: Fetch signing files
        script: |
          app-store-connect fetch-signing-files $BUNDLE_ID --type IOS_APP_STORE
      - name: Use system default keychain
        script: |
          keychain add-certificates
      - name: Set up code signing settings on Xcode project
        script: |
          xcode-project use-profiles
      - name: Increment build number
        script: |
          cd ios && agvtool new-version -all $(($(app-store-connect get-latest-testflight-build-number "$APP_STORE_APP_ID") + 1))
      - name: Build ipa for distribution
        script: |
          xcode-project build-ipa --project "$UNITY_IOS_DIR/$XCODE_PROJECT" --scheme "$XCODE_SCHEME"
    artifacts:
        - build/ios/ipa/*.ipa
        - $HOME/Library/Developer/Xcode/DerivedData/**/Build/**/*.dSYM
    publishing:
        app_store_connect:            
            api_key: $APP_STORE_CONNECT_PRIVATE_KEY         
            key_id: $APP_STORE_CONNECT_KEY_IDENTIFIER
            issuer_id: $APP_STORE_CONNECT_ISSUER_ID
    unity-android-workflow:
        name: Unity Android Workflow
        max_build_duration: 120
        environment:
            groups:

>             # Add the group environment variables in Codemagic UI (either in Application/Team variables) - https://docs.codemagic.io/variables/environment-variable-groups/

Reviewer: ```suggestion
            # Add the group environment variables in Codemagic UI (either in Application or Team variables) - https://docs.codemagic.io/variables/environment-variable-groups/
```
